### PR TITLE
Add DTS "soc-label" support

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

File: scripts/dts/extract/compatible.py [hunk 1]
```diff
@@ -43,6 +43,18 @@ def extract(self, node_address, prop, def_label):
                 compat_defs: "1",
             }
             insert_defs(node_address, load_defs, {})
+
+            # Generate defines of the form:
+            # #define DT_<COMPAT>_SOC_<SOC LABEL> 1
+            if 'soc-label' in reduced[node_address]['props'].keys():
```
Comment (ulfalizer) [inline scripts/dts/extract/compatible.py]: `x in dict` in Python always tests the keys, so could drop the `.keys()`.

Could do this too:

```python
soc_label = reduced[node_address]['props'].get('soc-label')
if soc_label is not None:
    soc_instance = ...
```

File: scripts/dts/extract/compatible.py [hunk 2]
```diff
@@ -43,6 +43,18 @@ def extract(self, node_address, prop, def_label):
...
+                soc_label = reduced[node_address]['props']['soc-label']
+
+                soc_instance = 'DT_' + compat_label + '_SOC_' + str(soc_label)
```
Comment (ulfalizer) [inline scripts/dts/extract/compatible.py]: Isn't `soc_label` already known to be a string?

File: scripts/dts/extract/compatible.py [hunk 3]
```diff
@@ -43,6 +43,18 @@ def extract(self, node_address, prop, def_label):
...
+                load_defs = {
+                    soc_instance: "1",
+                }
+                insert_defs(node_address, load_defs, {})
```
Comment (ulfalizer) [inline scripts/dts/extract/compatible.py]: Could do this maybe, to get rid of three lines:

```python
insert_defs(node_address, {soc_instance: "1"}, {})
```

**Review by ulfalizer — CHANGES_REQUESTED** [review]

**Review by erwango — COMMENTED** [review]
This raises me two questions:
- why not using complete soc doc label, ie 'UART_1' instead of '1' ? It's plausible devices names could vary from one SoC vendor to another (could be UART/USART for isntance)
-  I'm curious to see how devices instantiation would happen in driver code. We'll now have a mix of uncorrelated instances (starting from _0) defined at compile time and soc-label (alphabetical or numerical starting from _0 or _1) that are available depending on user's selection in device tree.

Taking STM32 example, with USART_2 being the first selected USART, here is what is generated if we include current PR (I'm focusing on few properties: base address, size, current-speed, label):
```
#define DT_ST_STM32_USART_0                      1
#define DT_ST_STM32_USART_40004400_BASE_ADDRESS	 0x40004400
#define DT_ST_STM32_USART_40004400_CURRENT_SPEED 115200
#define DT_ST_STM32_USART_40004400_LABEL	"UART_2"
#define DT_ST_STM32_USART_40004400_SIZE          1024
#define DT_ST_STM32_USART_SOC_2		         1
#define DT_ST_STM32_USART_0_BASE_ADDRESS	ST_ST_STM32_USART_40004400_BASE_ADDRESS
#define DT_ST_STM32_USART_0_CURRENT_SPEED	ST_ST_STM32_USART_40004400_CURRENT_SPEED
#define DT_ST_STM32_USART_0_LABEL		DT_ST_STM32_USART_40004400_LABEL
#define DT_ST_STM32_USART_0_SIZE		DT_ST_STM32_USART_40004400_SIZE
```

For this use case, USART_2 instantiation relies on DT_ST_STM32_USART_0_X props.
So DT_ST_STM32_USART_SOC_2 defined means a USART instance using DT_ST_STM32_USART_0_BASE_ADDRESS is available.
For another use case where USART_2 got isntance _1, DT_ST_STM32_USART_SOC_2 would be bounded with  DT_ST_STM32_USART_1_BASE_ADDRESS ...
Honestly I'm still not clear about intrications of this, and if I think about this in the good way. I might not. Maybe this is because I lack coffee, but definitely this is not quite intuitive and I guess some others could also be lost in wrong assumptions.

Though, since we commonly agreed that a vendor soc-label was required to identify devices instances for code generation, can't we use this to enumerate devices properties, so we don't have to rely on a compile time defined index which depends on the use case
So having DT_COMPAT_\<SOC_LABEL\>_\<PROP\> so we have DT_ST_STM32_USART_SOC_X in sync with DT_ST_STM32_USART_X_BASE_ADDRESS.

Then we can have the following in UART driver
```
#ifdef DT_ST_STM32_USART_SOC_X
static const struct uart_stm32_config uart_stm32_cfg_x = {		\
	.uconf = {							\
		.base = (u8_t *)DT_ST_STM32_USART_X_BASE_ADDRESS, 
[...]
#endif
```

**Comment by mnkp** [comment]
I believe this PR should be closed as it was superseded by the `DT_<COMPAT>_<ALIAS>_<PROP>` defines.

**Comment by erwango** [comment]
@galak , can we close this one?

**Pull request closed without merging**